### PR TITLE
[JENKINS-48365] Install detached plugins when upgrading Jenkins past the version the plugins were detached

### PR DESCRIPTION
**Pull request by dwnusbaum**
See [JENKINS-48365](https://issues.jenkins-ci.org/browse/JENKINS-48365).

[PluginManager](https://github.com/dwnusbaum/jenkins/blob/671a0ad4d073bb8622c16d3876c2f63a0c30171a/core/src/main/java/hudson/PluginManager.java#L681-L683) checks for `InstallState.UPGRADE` when deciding what logic to use. However, in Jenkins' constructor, the call to [Jenkins#executeReactor](https://github.com/jenkinsci/jenkins/blob/18d7ed5ac93f3f1d60ca887f49dbf78c4b8aaedb/core/src/main/java/jenkins/model/Jenkins.java#L904-L908), which will execute the 'Loading bundled plugins' task that calls `PluginManager#loadDetachedPlugins`, happens a few lines before the call to [InstallUtil#proceedToNext](https://github.com/jenkinsci/jenkins/blob/18d7ed5ac93f3f1d60ca887f49dbf78c4b8aaedb/core/src/main/java/jenkins/model/Jenkins.java#L927), which will update `InstallState` to UPGRADE if it detects that an upgraded occurred, so the upgrade branch is never taken in `PluginManager`. 

This PR instead checks the version numbers of the last running version and this version to decide if an upgrade occurred.

### Proposed changelog entries

* Make sure detached plugins (Plugins which used to be part of Jenkins itself) are installed when upgrading Jenkins past the version at which the plugin was detached.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees

### CODE REVIEW

File: test/src/test/java/jenkins/install/LoadDetachedPluginsTest.java
Comment (dwnusbaum) [inline test/src/test/java/jenkins/install/LoadDetachedPluginsTest.java]: This is a hack, but I don't think there is a way to test an actual upgrade with JTH (or ATH). Is the test worth keeping?

Comment (jglick) [reply test/src/test/java/jenkins/install/LoadDetachedPluginsTest.java]: Sure you can do it. Use `@LocalData`. The old version would be in `config.xml`. Not necessarily any better than what you have.

File: test/src/test/java/jenkins/install/LoadDetachedPluginsTest.java
Comment (jglick) [inline test/src/test/java/jenkins/install/LoadDetachedPluginsTest.java]: or

```java
assertThat("…", detachedPlugins, not(empty()));
```

**Review by jglick — APPROVED** [review]

File: test/src/test/java/jenkins/install/LoadDetachedPluginsTest.java
Comment (dwnusbaum) [reply test/src/test/java/jenkins/install/LoadDetachedPluginsTest.java]: Ah, that makes sense. I'll take a look at switching to local data because that should make it possible to test the edge cases in `InstallUtil#getLastExecVersions` by doing a pre 2.0 upgrade as well as a post 2.0 upgrade.

**Comment by daniel-beck** [comment]
Can we assert that all installed plugins loaded successfully to preempt e.g. dependency problems?

**Comment by dwnusbaum** [comment]
@daniel-beck That sounds like a good idea to me. Right now for detached plugins I assert `PluginWrapper#isActive` is true, and I will add an assertion that `PluginWrapper#getDependencyErrors` is empty, and an assertion that `PluginManager#getFailedPlugins` is empty. Is that what you had in mind?

**Comment by daniel-beck** [comment]
@dwnusbaum 

> Is that what you had in mind?

Exactly 👍

**Comment by recampbell** [comment]
@reviewbybees done

**Review by jglick — APPROVED** [review]

File: core/src/main/java/jenkins/install/InstallUtil.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/install/InstallUtil.java]: Make it restricted?

**Review by oleg-nenashev — COMMENTED** [review]
All recent issues in the installation logic just show how bad was our test suite for Jenkins 2.0 whet the Setup Wizard was introduced

I do not like that a constant is being made public without restrictions. It would be better to fix it before the merge

**Comment by daniel-beck** [comment]
Manual test looks reasonable.

	Dec 16, 2017 10:13:45 PM hudson.PluginManager loadDetachedPlugins
	INFO: Upgrading Jenkins. The last running version was 2.60.1. This Jenkins is version 2.94-SNAPSHOT (private-12/12/2017 22:40 GMT-win2012-deb6b0$).
	Dec 16, 2017 10:13:45 PM hudson.PluginManager loadDetachedPlugins
	INFO: Upgraded Jenkins from version 2.60.1 to version 2.94-SNAPSHOT (private-12/12/2017 22:40 GMT-win2012-deb6b0$). Loaded detached plugins (and dependencies): [command-launcher.hpi, script-security.hpi]

**Comment by daniel-beck** [comment]
@oleg-nenashev I pushed a commit to this PR that restricts the newly public field.

Will merge after PR build succeeds towards this weekly unless someone vetoes.

**Comment by dwnusbaum** [comment]
@oleg-nenashev @daniel-beck The `InstallUtil` class is already [@Restricted(NoExternalUse.class)](https://github.com/jenkinsci/jenkins/pull/3193/commits/2476d1417ecfbfb9864b86d544056421e67b0679#diff-8fce309eb0af96aa112306be6c389edbR67), is it necessary to restrictthe field as well?

**Review by oleg-nenashev — APPROVED** [review]
whatever works. This issue is pretty bad, so we need to integrate the change. 🐝 
@reviewbybees done

**Comment by daniel-beck** [comment]
@dwnusbaum Oops. I missed that. Good point. Oh well. We're going to merge today, so there will now be a pointless commit 🤷‍♂️

**Comment by daniel-beck** [comment]
Believed to have introduced [JENKINS-48604](https://issues.jenkins-ci.org/browse/JENKINS-48604) in 2.96.

**Pull request merged**
